arch/x86/chunkset_sse41.c

```diff
@@ -0,0 +1,103 @@
+/* chunkset_sse41.c -- SSE4 inline functions to copy small data chunks.
+ * For conditions of distribution and use, see copyright notice in zlib.h
+ */
+
+#include "zbuild.h"
+
+/* This requires SSE2 support. While it's implicit with SSE4, we can minimize
+ * code size by sharing the chunkcopy functions, which will certainly compile
+ * to identical machine code */
+#if defined(X86_SSE41) && defined(X86_SSE2)
```

Comment:
This doesn't really make sense... We can share the functions, but we don't need to test for `X86_SSE2`here... If user wants to disable SSE2 support, but not SSE41 support, then this will break horribly.

Reply:
As far as I can tell, there really isn't a cleaner way to dispatch the SSE2 functions, here.  You could go the long, roundabout way of dispatching through the function table but that has its own set of issues.  For one, you specifically want to call the shorter vector versions when you're on an architecture that has 256 bit wide versions, otherwise you're jumping twice (which maybe isn't the end of the world, but it'd be nice to avoid).  For two, you're losing a convenience afforded by having the macro named versions for each respective architecture when you want to call one from the other.  There's also the fact that calling into the function table does in fact require a trip to thread local storage, which does mean you have to visit segment registers to find the current address of your function (a small bit of overhead, but these are already routines dealing with relatively small chunks of data).

I think this is the lesser of two evils and it has no side effects as far as I've tested, even toggling all kinds of combinations of supported vectorizations into the build options.

Reply:
Recent GCC versions can't generate anything older than SSE4, so trying to compile code using just `-msse2` will most likely result in "bad" binary. I've worked with gcc people for over 6 years and some of my bug reports regarding older SSE versions are still open and unsolved. At work, we still use gcc 4.9 and gcc 6.5 as anything newer has issues that cause random crashes or poor performance depending on the CPU the code is run.

If someone wants to disable SSE2 code paths, it should revert to compiling the relevant functions using SSE4.1.

Reply:
If someone wants to disable SSE2 code paths, then we have no chunkcopy function, as the SSE4 version uses SSE2 chunkcopy directly, as does the SSE2 version.  So right now, disabling SSE2 code has the same effect.  If you don't have SSE2 support compiled in, you don't have accelerated chunkcopy, chunkmemset, etc (unless you happen to have AVX2).  

The SSE4 enabled paths for these same templates are going to compile to the exact same code, we are only increasing the binary size by doing this.  Right now (as in on develop), if you have an SSE4 enabled CPU and disable SSE2 support, you won't have vectorized chunk routines (exactly the same as in this PR).

Reply:
Here's chunkcopy_sse2 with only sse2 flags at build time:
```asm
0000000000013c40 <chunkcopy_sse2>:
   13c40:       8d 4a ff                lea    -0x1(%rdx),%ecx
   13c43:       f3 0f 6f 0e             movdqu (%rsi),%xmm1
   13c47:       83 e1 0f                and    $0xf,%ecx
   13c4a:       44 8d 41 01             lea    0x1(%rcx),%r8d
   13c4e:       0f 11 0f                movups %xmm1,(%rdi)
   13c51:       4a 8d 04 07             lea    (%rdi,%r8,1),%rax
   13c55:       4c 01 c6                add    %r8,%rsi
   13c58:       44 29 c2                sub    %r8d,%edx
   13c5b:       74 17                   je     13c74 <chunkcopy_sse2+0x34>
   13c5d:       01 c2                   add    %eax,%edx
   13c5f:       90                      nop
   13c60:       f3 0f 6f 06             movdqu (%rsi),%xmm0
   13c64:       48 83 c0 10             add    $0x10,%rax
   13c68:       48 83 c6 10             add    $0x10,%rsi
   13c6c:       0f 11 40 f0             movups %xmm0,-0x10(%rax)
   13c70:       39 c2                   cmp    %eax,%edx
   13c72:       75 ec                   jne    13c60 <chunkcopy_sse2+0x20>
   13c74:       c3                      ret    
   13c75:       66 66 2e 0f 1f 84 00    data16 cs nopw 0x0(%rax,%rax,1)
   13c7c:       00 00 00 00 

```

Here's with -msse4 applied to the global CFLAGS:

```asm
0000000000013c60 <chunkcopy_sse2>:
   13c60:       8d 4a ff                lea    -0x1(%rdx),%ecx
   13c63:       f3 0f 6f 0e             movdqu (%rsi),%xmm1
   13c67:       83 e1 0f                and    $0xf,%ecx
   13c6a:       44 8d 41 01             lea    0x1(%rcx),%r8d
   13c6e:       0f 11 0f                movups %xmm1,(%rdi)
   13c71:       4a 8d 04 07             lea    (%rdi,%r8,1),%rax
   13c75:       4c 01 c6                add    %r8,%rsi
   13c78:       44 29 c2                sub    %r8d,%edx
   13c7b:       74 17                   je     13c94 <chunkcopy_sse2+0x34>
   13c7d:       01 c2                   add    %eax,%edx
   13c7f:       90                      nop
   13c80:       f3 0f 6f 06             movdqu (%rsi),%xmm0
   13c84:       48 83 c0 10             add    $0x10,%rax
   13c88:       48 83 c6 10             add    $0x10,%rsi
   13c8c:       0f 11 40 f0             movups %xmm0,-0x10(%rax)
   13c90:       39 c2                   cmp    %eax,%edx
   13c92:       75 ec                   jne    13c80 <chunkcopy_sse2+0x20>
   13c94:       c3                      ret    
   13c95:       66 66 2e 0f 1f 84 00    data16 cs nopw 0x0(%rax,%rax,1)
   13c9c:       00 00 00 00 
```

The resulting machine code is basically identical.

Reply:
> If someone wants to disable SSE2 code paths, then we have no chunkcopy function, as the SSE4 version uses SSE2 chunkcopy directly, as does the SSE2 version. 

Obviously that is bad, but this pull request does nothing to make it more acceptable. We should in no case endorse use of incompatible or conflicting build options (except if building with compatibility enabled and zlib gracefully handles the build options). As such testing for `X86_SSE2` too early is unacceptable compromise.

Reply:
Honestly, I don't really see a big problem here.

Having duplicate functions would not be a good solution.

But an alternative that might be cleaner and more easily understood is to compile the required sse2 functions if `X86_SSE2` OR  `X86_SSE41` is defined (instead of only for sse2). That way it will always be available for either flag.
That would pull a subset of SSE2 functions into the SSE41 preference, but I think that only makes sense since there is no benefit in making a new SSE41 version of the specific functions. (Wholesale enabling all SSE2 functions would not make sense though).

Also these defines are not to be confused with the `-msse2` etc compiler flags, since those are very separate from this config-preference define.

Reply:
That `OR` is basically same as what I suggested. The real function gets compiled from `chunkset_sse2.c` if SSE2 is enabled, and from `chunkset_sse41.c` if SSE2 is disabled (for example to reduce binary size). This makes sure we always have optimized version if at least one vector instruction set is both enabled and supported.

Reply:
The way the templates are written that is impossible without a whole lot of nasty preprocessor churn to conditionally compile an SSE4 version if there is no sse2 version. It'd be ifdef spaghetti all for the same net effect.

SSE2 is a subset of SSE4 and it's actually a requirement now for 64 bit x86 processors. The odds of somebody trying to disable this on x86 are slim to none and misguided at best if they are optimization efforts.

Reply:
Rereading this, the suggestion for OR actually isn't as hard a pill to swallow as I was thinking had been suggested. I still contend it's a silly thing for the user to do, though. And, it's somewhat counter intuitive to what they might mean when disabling the WITH_SSE2 option.

Reply:
There is some merit to disabling the SSE2 functions if they can all be replaced by better alternatives AND the user knows that this will not run on any older cpus (or the C versions are ok for any stragglers).

This is kind of a gray area in our configuration options though, since it is currently hard to know whether you have all the "best" non-C functions enabled if you want to minimize the code size of the library. I guess it could be done in configure and cmake by having them print out each group of functions (adler, crc32, etc) along with the enabled instruction sets for each of them.

Reply:
The merit for that kind of dwindles when you realize that SSE4 doesn't add anything additional to the mix and doesn't justify creating SSE4 variants of them, though.  The memset function does, obviously, because we're leveraging shuffles and blends there, but all the other functions basically end up compiling to the same thing.  

On the flip side, renaming the copy functions to SSE4 and requiring SSE4 for it to work doesn't feel like a great solution, either.  Then older CPUs end up losing this capability.  Trying to make them mutually exclusive is messy.  We could split up the files as different compilation units but then you lose the benefit where the functions directly call each other (happens in memset and a couple of other routines) rather going through the indirection of the function table.  Not to mention the compiler also is more likely to inline these routines when they are called directly, so that has an added benefit for keeping them together.

Reply:
All I see is that if you move the test for `X86_SSE2` as late as possible, all the "ugly" magic ends up in single file, not "all over the sources"... I know it means the test has to be done multiple times (with fallback path) in this file for it to work as expected, but it's still better than having the test done in multiple files possibly unnecessarily.

Reply:
I say we modify the && to || in chunkset_tpl.h for this macro as suggested by @Dead2.  It doesn't make sense to name the other functions _sse4 as a fallback when sse2 support is not compiled in because the instructions they will generate will merely be comprised of the sse2 extensions with no additional instructions as far as I'm aware.  

The user turning off WITH_SSE2 also has other ill-effects that are plenty reason to motivate not doing it.  The slide hash functions, unless you have a CPU with AVX support, would also be disabled, as well as a myriad of other functions such as the fast compare routines which do not have faster counterparts until AVX support.  As of now, there are also a number of functions that are made faster by having both SSE support and AVX support compiled in (e.g. if you have fewer than 32 elements for an adler checksum and have the SSE functions compiled in, you'll avoid a lot of scalar calculations of that hash).  In benchmarks this was actually found to be quite a bit faster and I imagine it's commonplace to get small-ish windows like that for decompression.  I modified these things such that if they don't exist there's a safe fallback, but, if the user is looking to do this for speed optimization purposes, they're probably approaching it the wrong way.

Reply:
What I meant is that it would make sense in theory if a user knows that any machine they deploy to is for example a Haswell or newer cpu. In that case there is no upside to keeping around any functions that will never be used, such as SSE2 functions that have better suited versions.

We don't currently support properly selecting this though, but if we want to we could have a config parameters like `MIN_CPU=SSE4`  and`MAX_CPU=AVX2`  for example. That could then make sure we don't compile in AVX512 functions at all, as well as any SSE2/SSSE3 functions that have better SSSE3/SSE4 variants, unless needed by those replacements. It would still select between any SSE4/AVX1/AVX2 where available. If a function only has SSE2 and AVX512 implementations, then SSE2 would still be used since that is the best one available under those constraints. C fallbacks could potentially be kept or removed depending on a `C_FALLBACK` parameter for example.

This would of course mean a relatively major restructuring of CMakeLists.txt and probably some changes to defines used to select functions, but I think it does have some merit. The only major gain from this of course is that the library size is reduced and that can potentially improve cache locality.
A further step would be to bypass functable for functions where we know there is only the one option left, but that would probably be a lot more messy.

Reply:
I wouldn't suggest implementing `MIN_CPU` and `MAX_CPU` until we have is a use case and demand the justifies it, instead of just theory. It reminds me of all the work I did on the `attribute(target)` PR that never made it in.

Reply:
I get why the user might want to I just mean there are reasons that they shouldn't, mostly for the places where we do use both functions in tandem.

On my other PR I've been doing some interesting experiments.  I created a branch where this changed was merged into the inline adler changes and it became universally slower, but I think for reasons that are unrelated to either improvements.  The test case of the trivially compressible image for levels 7, 8, & 9 are still better off with this.  

I can't say for sure what's happening but I did notice that a lot of constants that are populated into vector registers did seem to be measuring some small amount of CPU time where they weren't before (and they were being loaded from a different base variable and offset from said base variable).  However, disabling SSE2 (and not chunkmemset_avx2), managed to restore a fair bit of performance.  This may be the effect that @Dead2 had been referring to with regard to function alignment, since while compiling fewer functions there, the binary wasn't appreciably smaller since we still had the permutation table.  

@Dead2 you mentioned something about moving hotter functions somewhere in the binary somehow?  Is this something PGO could maybe do?  Or maybe a custom linker script? You said you had a compile flag to at least make this consistent, what was it again?

Reply:
Hah, yeah, -falign-functions=32 seems to have brought back pretty much all of that performance.  *sigh*, compilers.  It is interesting how little function alignment is discussed in the realm of optimization, though.  Even PGO didn't seem to pick up on this.

Interesting reads:
https://easyperf.net/blog/2018/01/25/Code_alignment_options_in_llvm
https://easyperf.net/blog/2018/01/18/Code_alignment_issues

Reply:
Yes, function alignment is part of the "problem" since it is aligned to 16 bytes by default (at least on x86-64) while cachelines are 64 bytes. So when I do comparison benchmarks I add this CMake parameter `-DCMAKE_C_FLAGS="-falign-functions=64"`. This means that each function is always placed the same in the cacheline across multiple builds, instead of starting at a random 16-byte aligned spot (sometimes this makes a big difference when a hot function switches between crossing cachelines and not).

The other problem is cache aliasing, and that is more likely to get worse with the above, since the change above means we request more cachelines total (max 1 function per cacheline instead of max 4 functions), so there is a bigger chance of cachelines getting demoted to L2 for example. But I don't think our code size is big enough that this usually matters for most L1I caches.

In release builds you can get some performance improvements by having multiple related-to-each-other functions in the same cacheline.

PGO utilizes this effect by grouping hot functions so that no cold functions end up residing in the same cacheline as hot functions. This relies on good profiling however, and since we have so many exported functions it might be hit/miss depending on actual use. A more manual approach (guided by profiling) would be to add gcc/clang attributes that group certain functions into separate blocks, such as "inflate", "deflate", "common" and the rest of the more cold functions into the default block. This might be less of an improvement than PGO (since PGO also does optimization changes for hot/cold functions), but unlike PGO it would not mean any extra steps for the packager/end-user.

Regarding MIN_CPU/MAX_CPU, that would be a pretty neat thing to support, and probably not very hard either, but it would of course require that situations where for example SSE4 functions requiring certain SSE2 functions actually pulls those in automatically as well instead of being a strict lower exclusion limit.

😄

Reply:
When you specify minimum CPU high enough, compiler can optimize the code better as the related functions are all in same compilation unit. This is kind of half-way between having LTO disabled and enabled, as some function variants still need to make extern calls to "shorter" variants. This is similar to bypassing functable when there is only one or two versions (basically unoptimized C variant and one vectorized variant).

Reply:
From my limited testing in just trying to use PGO on the benchmark build artifacts (which is of course, probably stupid but a fun experiment), PGO wasn't able to pick function alignments nearly as well as the static alignment specified to GCC.  It certainly seems like there are a lot of functions we know will be cold (all of the stub functions, the cpu detection functions, etc), so I do wonder if a custom linker script could group all of those together to get us a tighter binary while allowing us to put the hotter functions with loops where alignment really matters to the instruction cache at 32 or 64 byte aligned boundaries.  Then again, a `#pragma GCC optimize()` macro could probably give use a similar effect.

There also might be some fruit in reordering some of the fields in these structs such as inflate_state.  I know things accessed frequently together should be close together but there might be some amount of padding that is occurring between struct members as well for alignment purposes (not egregious padding, since everything is mostly 32 or 64 bit, but some).  For sure the crc_fold struct should probably be first, since its first members require 64 byte alignment.  The "work" and "len" structs, while a modulus of 4 bytes, might still be better at the tail end of the struct.  The "dmax" field probably also can be guarded by a macro conditionally if INFLATE_STRICT is enabled or not.

Reply:
@KungFuJesus The attribute I mentioned earlier for grouping inflate/deflate/common etc is called `section` :
https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html

Separating inflate and deflate functions should provide better grouping than just bunching them all into one big "warm" section. Of course that also requires a common section for any shared hot functions, but I don't imagine having a couple hot sections switch between incurs as big a penalty as just having random hot/cold functions bunched together.
All the cold code should then just go to the generic `text` section like it normally would.
Using these should also be easier than using and maintaining linker-scripts.
There is a risk that these could get stripped if a packager is too eager to strip any "extra" data from the library/executable though, but then it would also quickly fail any actual test.

While the attributes `hot` and `cold` also put code into different sections, these affect optimization as well and also cause branch prediction changes. `cold` should really only be used for functions that nearly never get used, such as error handlers. These are the ones used by PGO, but carry a higher risk for slowing down the code when done manually, also the above illustrates the need for good and accurate profiling for PGO (something more easily done for a specific applications use-case rather than library-wide).

Reply:
Interesting, I've never seen these gcc attributes before. The PMCs listed in that article are available on my CPU, it might be an interesting experiment to find the hotter icache miss stalls from that.